client/components/dashboard/dashboard-service.js

```diff
@@ -547,34 +548,34 @@ DashboardService.prototype.rewriteQuery = function(widget, replaceParams) {
 
   let widgetConfig = widget.model.datasource.config;
 
-  let project_name = this.arrayUtilService_.getFirst([
+  let project_name = /** @type {string} */ (this.arrayUtilService_.getFirst([
       widgetConfig.results.project_id,
       this.current.model.project_id,
-      this.config.default_project], false);
+      this.config.default_project], false));
   if (project_name === null) {
     this.errorService_.addError(ErrorTypes.DANGER, 'Project name not found.');
   }
 
-  let dataset_name = this.arrayUtilService_.getFirst([
+  let dataset_name = /** @type {string} */ (this.arrayUtilService_.getFirst([
       widgetConfig.results.dataset_name,
       this.current.model.dataset_name,
-      this.config.default_dataset], false);
+      this.config.default_dataset], false));
   if (project_name === null) {
     this.errorService_.addError(ErrorTypes.DANGER, 'Dataset name not found.');
   }
 
-  let table_name = this.arrayUtilService_.getFirst([
+  let table_name = /** @type {string} */ (this.arrayUtilService_.getFirst([
       widgetConfig.results.table_name,
       this.current.model.table_name,
-      this.config.default_table], false);
+      this.config.default_table], false));
   if (project_name === null) {
     this.errorService_.addError(ErrorTypes.DANGER, 'Table name not found.');
   }
 
-  let table_partition = this.arrayUtilService_.getFirst([
+  let table_partition = /** @type {!QueryTablePartitioning} */ (this.arrayUtilService_.getFirst([
```

Comment:
Also added a note to the ArrayUtilService to refactor it into a non-angular library, and use Generics to handle casting.